File: flutter_app/lib/models/task.dart
class Task {
  final int id;
  final String title;
  final String? description;
  final String taskType;
  final DateTime? dueDate;
  final String priority;
  final String status;
  final bool overdue;
  final DateTime? completedAt;
  final int userId;
  final Map<String, dynamic>? createdBy;
  final int? scheduleEntryId;

  Task({
    required this.id,
    required this.title,
    this.description,
    required this.taskType,
    this.dueDate,
    required this.priority,
    required this.status,
    required this.overdue,
    this.completedAt,
    required this.userId,
    this.createdBy,
    this.scheduleEntryId,
  });

  bool get isCompleted => status == 'completed';
  bool get isCancelled => status == 'cancelled';
  bool get isActive => !isCompleted && !isCancelled;

  factory Task.fromJson(Map<String, dynamic> json) {
    return Task(
      id: json['id'],
      title: json['title'],
      description: json['description'],
      taskType: json['task_type'] ?? 'general',
      dueDate: json['due_date'] != null ? DateTime.parse(json['due_date']) : null,
      priority: json['priority'] ?? 'medium',
      status: json['status'] ?? 'pending',
      overdue: json['overdue'] ?? false,
      completedAt: json['completed_at'] != null ? DateTime.parse(json['completed_at']) : null,
      userId: json['user_id'],
      createdBy: json['created_by'],
      scheduleEntryId: json['schedule_entry_id'],
    );
  }
}

File: app/models/task.rb
class Task < ApplicationRecord
  include Auditable

  TASK_TYPES = %w[submission_deadline preparation_reminder follow_up checklist general].freeze
  STATUSES = %w[pending in_progress completed cancelled].freeze
  PRIORITIES = %w[low medium high urgent].freeze

  belongs_to :user
  belongs_to :created_by, class_name: "User"
  belongs_to :schedule_entry, optional: true

  has_many :notifications, as: :notifiable, dependent: :destroy

  validates :title, presence: true
  validates :task_type, inclusion: { in: TASK_TYPES }
  validates :status, inclusion: { in: STATUSES }
  validates :priority, inclusion: { in: PRIORITIES }

  before_save :set_completed_at

  scope :pending, -> { where(status: "pending") }
  scope :active, -> { where(status: %w[pending in_progress]) }
  scope :overdue, -> { active.where("due_date < ?", Date.today) }
  scope :due_today, -> { active.where(due_date: Date.today) }
  scope :due_this_week, -> { active.where(due_date: Date.today..Date.today + 7.days) }

  def complete!(actor = nil)
    update!(status: "completed", completed_at: Time.current)
  end

  def overdue?
    due_date.present? && due_date < Date.today && %w[pending in_progress].include?(status)
  end

  private

  def set_completed_at
    self.completed_at = Time.current if status_changed? && status == "completed" && completed_at.nil?
  end
end
